src/layers/activations/softmax.cu

```diff
@@ -101,5 +196,70 @@ void bp_cutoff(int height, int width,
     cutoff);
 }
 
+void max_local_col_entry(int height, int width,
+                         const DataType * __restrict__ input,
+                         int input_ldim,
+                         DataType * __restrict__ workspace,
+                         cudaStream_t stream) {
+  if (width <= 0 || height <= 0) {
+    return;
+  }
+  const int block_dim = 256;
+  const int grid_dim = (width + block_dim - 1) / block_dim;
+  max_local_col_entry_kernel<<<grid_dim, block_dim, 0, stream>>>(
+    height, width, input, input_ldim, workspace);
+}
+
+void exp_and_col_sum(int height, int width,
+                     const DataType * __restrict__ input,
+                     int input_ldim,
+                     DataType * __restrict__ output,
+                     int output_ldim,
+                     DataType * __restrict__ workspace,
+                     cudaStream_t stream) {
+  if (width <= 0 || height <= 0) {
```

Comment:
`workspace` should be set to zero when `width > 0` and `height == 0`.

Reply:
Done.